src/constraint.rs

```diff
@@ -1550,35 +1530,47 @@ impl ValidValuesConstraint {
 }
 
 impl ConstraintValidator for ValidValuesConstraint {
-    fn validate(&self, value: &OwnedElement, type_store: &TypeStore) -> ValidationResult {
-        for valid_value in &self.valid_values {
-            match valid_value {
-                ValidValue::Range(range) => match value.ion_type() {
-                    IonType::Integer | IonType::Float | IonType::Decimal | IonType::Timestamp => {
-                        if range.contains(value) {
-                            return Ok(());
+    fn validate(&self, value: &IonSchemaElement, type_store: &TypeStore) -> ValidationResult {
+        match value {
+            IonSchemaElement::SingleElement(value) => {
+                for valid_value in &self.valid_values {
+                    match valid_value {
+                        ValidValue::Range(range) => match value.ion_type() {
+                            IonType::Integer
+                            | IonType::Float
+                            | IonType::Decimal
+                            | IonType::Timestamp => {
+                                if range.contains(value) {
+                                    return Ok(());
+                                }
+                            }
+                            _ => {}
+                        },
+                        ValidValue::Element(element) => {
+                            // get value without annotations
+                            let value = value.to_owned().with_annotations(vec![]);
+
+                            if element == &value {
+                                return Ok(());
+                            }
                         }
-                    }
-                    _ => {}
-                },
-                ValidValue::Element(element) => {
-                    // get value without annotations
-                    let value = value.to_owned().with_annotations(vec![]);
-
-                    if element == &value {
-                        return Ok(());
-                    }
+                    };
                 }
-            };
+                Err(Violation::new(
+                    "valid_values",
+                    ViolationCode::InvalidValue,
+                    &format!(
+                        "expected valid_values to be from {:?}, found {:?}",
+                        &self, value
+                    ),
+                ))
+            }
+            IonSchemaElement::Document(document) => Err(Violation::new(
+                "valid_values",
+                ViolationCode::InvalidValue,
+                "valid_values constraint is not allowed to be used for document",
```

Comment:
minor suggestion—rather than having a different message (and having people wonder "why can't it be used for a document"), you could use the same message. `valid_values` will never match a `document` because there's no way to create a `valid_values` constraint with a `document` in it.

Reply:
Are you saying that the violation message should just say:

> `valid_values` will never match a document because there's no way to create a `valid_values` constraint with a document in it.

Reply:
I'm suggesting it should be:
```rust
&format!("expected valid_values to be from {:?}, found {:?}", &self, value)
```

Reply:
Oh okay, got it.